### PR TITLE
Hybridize CoverCrypt using Kyber

### PR DESCRIPTION
**Pull request by tbrezot**
In this solution, keys and encapsulations contain sub-keys or sub-encapsulations that can be either classic of hybridized.

**cons**:
    - 1 bybte per partition is added to all serializations;
    - hybridized sub-encapsulations can be spotted in a encapsulation because their size differs from the classic ones;

**pros**:
    - hybridization can be used on a subset of partitions only;
    - code is leaner;

CoverCrypt now uses the `Policy` object to determine where to use hybridized encryption. This allows to keep a lean API while having a fine grained control over the encryption settings.

### CODE REVIEW

**Comment by tbrezot** [comment]
Benchmark need to be updated.

File: src/core/partitions.rs [hunk 1]
```diff
@@ -67,217 +64,158 @@ impl From<&[u8]> for Partition {
...
-/// corresponding list of `CoverCrypt` partitions; this only gets the current
-/// partitions, not the old ones
+/// Converts a list of attributes into the list of current `Partitions`, with
+/// their associated hybridization hints.
 ///
 /// - `attributes`  : list of attributes
-/// - `policy`      : security policy
-pub(crate) fn to_partitions(
+/// - `policy`      : global policy data
+fn generate_current_attriubte_partitions(
```
Comment (ThibaudDauce) [inline src/core/partitions.rs:108]:
```suggestion
fn generate_current_attribute_partitions(
```

File: src/core/partitions.rs [hunk 2]
```diff
@@ -67,217 +64,158 @@ impl From<&[u8]> for Partition {
...
+    // Recursive call. Above checks ensure no empty list can be returned.
+    let other_values = combine_attribute_values(current_axis + 1, axes, attr_values_per_axis)?;
+
+    let mut combinations = Vec::with_capacity(current_axis_values.len() * other_values.len());
+    for (current_values, is_hybridized) in current_axis_values {
+        for (other_values, is_other_hybridized) in &other_values {
+            let mut combined = Vec::with_capacity(1 + other_values.len());
+            combined.push(*current_values);
+            combined.extend_from_slice(other_values);
+            combinations.push((combined, *is_hybridized | *is_other_hybridized));
```
Comment (ThibaudDauce) [inline src/core/partitions.rs:185]: I didn't know the `|`  operator, isn't it clearer to use normal OR here `||`? It seems to be the same for `bool` but maybe I'm missing something (`||` seems more correct for me here)

File: src/core/partitions.rs [hunk 3]
```diff
@@ -67,217 +64,158 @@ impl From<&[u8]> for Partition {
...
 ) -> Result<HashSet<Partition>, Error> {
     let attr_combinations =
-        to_attribute_combinations(access_policy, policy, follow_hierarchical_axes)?;
-    let mut set = HashSet::with_capacity(attr_combinations.len());
+        access_policy.to_attribute_combinations(policy, follow_hierarchical_axes)?;
+    let mut res = HashSet::with_capacity(attr_combinations.len());
     for attr_combination in &attr_combinations {
-        for partition in to_partitions(attr_combination, policy)? {
-            let is_unique = set.insert(partition);
+        for partition in generate_current_attriubte_partitions(attr_combination, policy)? {
```
Comment (ThibaudDauce) [inline src/core/partitions.rs:205]:
```suggestion
        for partition in generate_current_attribute_partitions(attr_combination, policy)? {
```

File: src/core/partitions.rs [hunk 4]
```diff
@@ -340,11 +278,11 @@ mod tests {
         )?;
         assert_eq!(partitions_1.len(), 1);
         let att_1_0 = axes_attributes[1][0].1;
-        assert!(partitions_1.contains(&Partition::from_attributes(vec![att_0_0, att_1_0])?));
+        assert!(partitions_1.contains(&Partition::from_attribute_values(vec![att_0_0, att_1_0])?));
 
         // this should create the 2 combinations of the first attribute
         // of the first axis with that the wo of the second axis
-        let partitions_2 = to_partitions(
+        let partitions_2 = generate_current_attriubte_partitions(
```
Comment (ThibaudDauce) [inline src/core/partitions.rs:285]:
```suggestion
        let partitions_2 = generate_current_attribute_partitions(
```

File: src/core/partitions.rs [hunk 5]
```diff
@@ -355,16 +293,16 @@ mod tests {
...
+        assert!(partitions_2.contains(&Partition::from_attribute_values(vec![att_0_0, att_1_1])?,));
 
         // rotation
         policy.rotate(&axes_attributes[0][0].0)?;
         let axes_attributes = axes_attributes_from_policy(&axes, &policy)?;
 
         // this should create the single combination of the first attribute
         // of the first axis with that of the second axis
-        let partitions_3 = to_partitions(
+        let partitions_3 = generate_current_attriubte_partitions(
```
Comment (ThibaudDauce) [inline src/core/partitions.rs:305]:
```suggestion
        let partitions_3 = generate_current_attribute_partitions(
```

File: src/core/primitives.rs [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+        assert!(res0.is_err(), "User 0 shouldn't be able to decapsulate!");
+
+        let res1 = decaps!(&usk1, &encapsulation)?;
+
+        assert_eq!(sym_key, res1, "Wrong decapsulation for user 1!");
+
+        // Change partition
+        let client_partition = Partition(b"client".to_vec());
+        let new_partitions_set = HashMap::from([
+            (dev_partition, EncryptionHint::Classic),
```
Comment (ThibaudDauce) [inline src/core/primitives.rs:488]: Could you try upgrading a partition from classic to hybrid / hybrid to classic in a rotate? Not sure what is suppose to be the rule here but we should have a test to check that it's not something wrong?

File: src/error.rs [hunk 7]
```diff
@@ -3,14 +3,16 @@
...
+use pqc_kyber::KyberError;
 use thiserror::Error;
 
 #[derive(Error, Debug)]
 pub enum Error {
     #[error("Unknown partition {0}")]
     UnknownPartition(String),
     #[error("{0}")]
-    CryptoError(CryptoCoreError),
+    CryptoError(String),
```
Comment (ThibaudDauce) [inline src/error.rs:15]: Could we keep the typed `CryptoCoreError` error here?

File: src/core/primitives.rs (hunk 6 above)
Comment (tbrezot) [reply src/core/primitives.rs:488]: Sure, i thought i did something like that but i might have slipped  throught

File: src/error.rs (hunk 7 above)
Comment (tbrezot) [reply src/error.rs:15]: ok

**Comment by tbrezot** [comment]
> Benchmarks need to be updated.

Benchmarks have been updated in #88.

**Review by Manuthor — APPROVED** [review]

**Pull request merged**